src/Generics/SOP/DMapUtilities.hs

```diff
@@ -0,0 +1,174 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE ExplicitNamespaces #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+{-|
+Module           : Generics.SOP.DMapUtilities
+Description      : Utilities for converting between the NS/NP types of generics-sop and Dependent Maps.
+-}
+
+module Generics.SOP.DMapUtilities
+  (
+    -- * Type Functions
+    FunctorWrapTypeList
+  , FunctorWrapTypeListOfLists
+
+    -- * Types
+  , TypeListTag
+
+    -- * Conversions
+    -- ** 'NP' \<-\> 'DM.DMap'
+  , npToDMap
+  , dMapToNP
+    -- ** 'NS' \<-\> 'DSum'
+  , nsToDSum
+  , dSumToNS
+
+  -- * Functor wrapping/unwrapping utilities for 'NP'
+  , npUnCompose
+  , npReCompose
+  , nsOfnpReCompose
+
+  -- * Utilities
+  , npSequenceViaDMap
+
+  -- * Proofs
+  , functorWrappedSListIsSList
+  )where
+
+import Generics.SOP ((:.:) (Comp), K (..), NP (..), NS (..), Proxy (..), SList (..), SListI (..), SListI2, fn,
+                     hcollapse, hmap, unComp)
+import Generics.SOP.Dict (Dict (Dict), withDict)
+import Generics.SOP.NP (sequence'_NP)
+import Generics.SOP.NS (ap_NS)
+
+import qualified Data.Dependent.Map as DM
+import Data.Dependent.Sum (DSum ((:=>)))
+import qualified Data.Dependent.Sum as DS
+
+import Data.GADT.Compare ((:~:) (..), GCompare (..), GEq (..), GOrdering (..))
+
+
+import Data.Functor.Identity (Identity (runIdentity))
+
+
+-- |A Tag type for making a 'DM.DMap' keyed by a type-level list
+data TypeListTag (xs :: [k]) (x :: k) where -- x is in xs
+  Here  :: TypeListTag (x ': xs) x          -- x begins xs
+  There :: TypeListTag xs x -> TypeListTag (y ': xs) x -- given that x is in xs, x is also in (y ': xs)
```

Comment:
`Here` and `There` are very nearly a naming conflict with `here` and `there` in `Data.These`.  Are these names chosen to be consistent with something else, or would it be possible to change them to avoid this near-conflict?